src/librbd/image/CreateRequest.cc

```diff
@@ -0,0 +1,696 @@
+  m_op_work_queue(op_work_queue), m_on_finish(on_finish) {
+
+  m_ioctx.dup(ioctx);
+  m_cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+
+  m_id_obj = util::id_obj_name(m_imgname);
+  m_header_obj = util::header_name(m_imageid);
+  m_objmap_name = ObjectMap::object_map_name(m_imageid, CEPH_NOSNAP);
+
+  m_layout = file_layout_t();
+  m_layout.object_size = 1ull << m_order;
+  if (m_stripe_unit == 0 || m_stripe_count == 0) {
+    m_layout.stripe_unit = m_layout.object_size;
+    m_layout.stripe_count = 1;
+  } else {
+    m_layout.stripe_unit = m_stripe_unit;
+    m_layout.stripe_count = m_stripe_count;
+  }
+
+  m_force_non_primary = !non_primary_global_image_id.empty();
+}
+
+template<typename I>
+CreateRequest<I>::~CreateRequest() {
+  m_ioctx.close();
+}
+
+template<typename I>
+void CreateRequest<I>::send() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  if (!validate_features(m_cct, m_features, m_force_non_primary)) {
+    complete(-EINVAL);
+    return;
+  }
+  if (!validate_layout(m_cct, m_size, m_layout)) {
+    complete(-EINVAL);
+    return;
+  }
+
+  validate_pool();
+}
+
+template<typename I>
+void CreateRequest<I>::validate_pool() {
+  if (!m_cct->_conf->rbd_validate_pool) {
+    create_id_object();
+    return;
+  }
+
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  using klass = CreateRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_validate_pool>(this);
+
+  librados::ObjectReadOperation op;
+  op.stat(NULL, NULL, NULL);
+
+  int r = m_ioctx.aio_operate(RBD_DIRECTORY, comp, &op, &m_outbl);
+  assert(r == 0);
+  comp->release();
+}
+
+template<typename I>
+Context* CreateRequest<I>::handle_validate_pool(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  if (*result == 0) {
+    create_id_object();
+    return nullptr;
+  } else if ((*result < 0) && (*result != -ENOENT)) {
+    lderr(m_cct) << "failed to stat RBD directory: " << cpp_strerror(*result) << dendl;
+    return m_on_finish;
+  }
+
+  // allocate a self-managed snapshot id if this a new pool to force
+  // self-managed snapshot mode
+  // This call is executed just once per (fresh) pool, hence we do not
+  // try hard to make it asynchronous (and it's pretty safe not to cause
+  // deadlocks).
+
+  uint64_t snap_id;
+  int r = m_ioctx.selfmanaged_snap_create(&snap_id);
+  if (r == -EINVAL) {
+    lderr(m_cct) << "pool not configured for self-managed RBD snapshot support"
+                 << dendl;
+    *result = r;
+    return m_on_finish;
+  } else if (r < 0) {
+    lderr(m_cct) << "failed to allocate self-managed snapshot: "
+                 << cpp_strerror(r) << dendl;
+    *result = r;
+    return m_on_finish;
+  }
+
+  r = m_ioctx.selfmanaged_snap_remove(snap_id);
+  if (r < 0) {
+    lderr(m_cct) << "failed to release self-managed snapshot " << snap_id
+                 << ": " << cpp_strerror(r) << dendl;
+    *result = r;
```

Comment:
Minor: ignore the error -- no recovery so no point in failing since we have accomplished the purpose of switching to self-managed snapshot mode on the pool.